consensus/abci.go

```diff
@@ -97,10 +99,24 @@ func (abci *TendermintABCI) InitChain(req tmtAbciTypes.RequestInitChain) tmtAbci
 // for the validators.
 //
 // Response:
-//		- Optional Key-Value tags for filtering and indexing
+// 		- Optional Key-Value tags for filtering and indexing
 func (abci *TendermintABCI) BeginBlock(req tmtAbciTypes.RequestBeginBlock) tmtAbciTypes.ResponseBeginBlock {
-	abci.logger.Debug("Beginning new block", "hash", req.Hash)
-	abci.db.UpdateBlockState(&req.Header)
+	abci.logger.Info("Beginning new block", "hash", tmtHexHash(req.Hash), "height", req.Header.Height, "time", req.Header.Time.String())
+	parentBlock := abci.getCurrentBlock()
+
+	// IMPORTANT: According to Tendermint documentation and based on the consensus setup made for our network
+	// two consecutive blocks cannot generated less than one second apart. BUT we identify an issue on that assumption
+	// and reported here https://github.com/tendermint/tendermint/issues/3755. Therefore we implemented this mitigation code
+	// to prevent ethereum headers to be invalid. Learn more in https://github.com/lightstreams-network/lightchain/issues/186
+	if uint64(req.Header.Time.Unix()) <= parentBlock.Time() {
+		abci.metrics.ReplacedBlockTimeTotal.Add(1)
+		nextBlockTime := time.Unix(int64(parentBlock.Time() + 1), 0)
```

Comment:
I will change this to `validEthBlockTime` to avoid confusing with actual "next" "block".